ClosedXML/Excel/Cells/XLCell.cs

```diff
@@ -218,57 +218,125 @@ internal IXLCell SetValue<T>(T value, bool setTableHeader)
                     Style.Alignment.WrapText = true;
+
+                parsed = true;
             }
-            else if (value is TimeSpan)
+            else
             {
-                _cellValue = value.ToString();
-                _dataType = XLDataType.TimeSpan;
-                if (style.NumberFormat.Format == String.Empty && style.NumberFormat.NumberFormatId == 0)
-                    Style.NumberFormat.NumberFormatId = 46;
+                var tuple = SetKnownTypedValue(value, style);
+                parsedValue = tuple.Item1;
+                parsed = tuple.Item2;
             }
-            else if (value is DateTime)
-            {
-                _dataType = XLDataType.DateTime;
-                var dtTest = (DateTime)Convert.ChangeType(value, typeof(DateTime));
-                if (style.NumberFormat.Format == String.Empty && style.NumberFormat.NumberFormatId == 0)
-                    Style.NumberFormat.NumberFormatId = dtTest.Date == dtTest ? 14 : 22;
 
-                _cellValue = dtTest.ToOADate().ToInvariantString();
+            // If parsing was unsuccessful, we throw an ArgumentException
+            // because we are using SetValue<T> (typed).
+            // Only in SetValue(object value) to we try to fall back to a value of a different type
+            if (!parsed)
+                throw new ArgumentException($"Unable to set cell value to {value.ToInvariantString()}", nameof(value));
+
+            SetInternalCellValueString(parsedValue, validate: true, parseToCachedValue: false);
+
+            CachedValue = null;
+
+            return this;
+        }
+
+        private Tuple<string, bool> SetKnownTypedValue<T>(T value, XLStyleValue style)
+        {
+            string parsedValue;
+            bool parsed;
+            if (value is DateTime d && d >= BaseDate)
+            {
+                parsedValue = d.ToOADate().ToInvariantString();
+                parsed = true;
+                SetDateTimeFormat(style, d.Date == d);
             }
-            else if (value.GetType().IsNumber())
+            else if (value is TimeSpan ts)
             {
-                if ((value is double || value is float) && (Double.IsNaN((Double)Convert.ChangeType(value, typeof(Double)))
-                    || Double.IsInfinity((Double)Convert.ChangeType(value, typeof(Double)))))
+                parsedValue = ts.TotalDays.ToInvariantString();
+                parsed = true;
+                SetTimeSpanFormat(style);
+            }
+            else if (value is Boolean b)
+            {
+                parsedValue = b ? "1" : "0";
+                _dataType = XLDataType.Boolean;
+                parsed = true;
+            }
+            else if (value.IsNumber())
+            {
+                if ((value is double || value is float)
+                    && (Double.IsNaN((Double)Convert.ChangeType(value, typeof(Double)))
+                        || Double.IsInfinity((Double)Convert.ChangeType(value, typeof(Double)))))
```

Comment:
```
if ((value is double d && (double.IsNaN(d) || double.IsInfinity(d)) ||
    (value is float f && (float.IsNaN(f) || float.IsInfinity(f))
```
looks cleaner and does not require `(Double)Convert.ChangeType(value, typeof(Double))` to be executed twice.

Reply:
True. And I must add a test for this.

Reply:
done.